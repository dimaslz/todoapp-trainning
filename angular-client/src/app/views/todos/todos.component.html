<div class="w-full">
  <header class="Header flex py-5 px-20 w-full">
    <h1
      class="
        w-full
        flex
        items-center
        justify-center
        text-2xl text-red-600
        font-extrabold
        uppercase
      "
    >
      TODO's angular app
    </h1>
  </header>

  <div class="px-4 mt-12">
    <div class="flex py-2 border-b items-center">
      <form
        [formGroup]="form"
        (ngSubmit)="onSubmit()"
        novalidate
        class="flex items-center justify-center h-full w-full"
      >
        <div class="flex h-full w-full">
          <input
            type="text"
            name="title"
            formControlName="title"
            class="
              border
              w-full
              h-10
              flex flex-grow
              outline-none
              p-2
              text-base
              focus:ring-1
            "
          />
        </div>

        <div class="flex h-full">
          <cui-button type="submit" size="large" [disabled]="!canCreate"
            >create</cui-button
          >
        </div>
      </form>
    </div>

    <div class="w-full h-full relative mt-6">
      <div
        *ngIf="loading; else todosContainer"
        class="w-full h-full absolute inset-0"
      >
        loading
      </div>

      <ng-template #todosContainer>
        <div
          class="flex flex-col w-full items-center justify-center"
          *ngIf="!todos.length; else todosLoaded"
        >
          <div>
            <check-icon class="text-green-500" [size]="150"></check-icon>
          </div>
          <div class="text-xl mt-4">TODO is empty!</div>
        </div>

        <ng-template #todosLoaded>
          <div
            *ngFor="let todo of todos; trackBy: getTodoID"
            class="border-b p-2 flex"
          >
            <div class="flex flex-col flex-grow">
              <h2 class="Todo__title flex w-full">
                <document-icon class="text-gray-400"></document-icon>
                <div class="ml-4 text-gray-800">{{ todo.title }}</div>
              </h2>
              <div
                *ngIf="todo.description"
                class="Todo__description text-sm text-gray-600"
              >
                ...
              </div>
            </div>
            <div class="flex">
              <cui-button
                [circle]="true"
                [outline]="true"
                (click)="deleteTodo(todo._id)"
                color="danger"
              >
                <trash-icon [size]="18"></trash-icon
              ></cui-button>
            </div>
          </div>
        </ng-template>
      </ng-template>
    </div>
  </div>
</div>
