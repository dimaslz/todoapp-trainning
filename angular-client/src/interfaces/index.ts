export { Todo, UpdateTodo, CreateTodo } from "./todo.interface";
