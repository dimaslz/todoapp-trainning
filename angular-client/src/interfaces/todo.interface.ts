export type Todo = {
  _id: string;
  title: string;
  description: string;
  status: {
    _id: string;
    name: string;
  };
};

export type CreateTodo = {
  title?: string;
  description?: string;
  status?: string;
};

export type UpdateTodo = {
  title?: string;
  description?: string;
  status?: string;
};
