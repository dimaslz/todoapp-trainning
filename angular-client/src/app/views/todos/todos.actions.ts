import { CreateTodo, Todo, UpdateTodo } from "@/interfaces";
import { createAction, props } from "@ngrx/store";

export const GET_TODOS = '[Todos] Get todos';
export const GET_TODOS_SUCCESS = '[Todos] Get todos success';
export const GET_TODOS_FAILED = '[Todos] Get todos failed';

export const CREATE_TODO = '[Todos] Create todo';
export const CREATE_TODO_SUCCESS = '[Todos] Create todo success';
export const CREATE_TODO_FAILED = '[Todos] Create todo failed';

export const UPDATE_TODO = '[Todos] Update todo';
export const UPDATE_TODO_SUCCESS = '[Todos] Update todo success';
export const UPDATE_TODO_FAILED = '[Todos] Update todo failed';

export const DELETE_TODO = '[Todos] Delete todo';
export const DELETE_TODO_SUCCESS = '[Todos] Delete todo success';
export const DELETE_TODO_FAILED = '[Todos] Delete todo failed';

export const getTodos = createAction(
  GET_TODOS
);

export const getTodosSuccess = createAction(
  GET_TODOS_SUCCESS,
  props<{
		data: Todo[];
	}>()
);

export const getTodosFailed = createAction(
  GET_TODOS_FAILED,
  props<{
		error: string;
	}>()
);

export const createTodo = createAction(
  CREATE_TODO,
  props<{
		todo: CreateTodo;
	}>()
);

export const createTodoSuccess = createAction(
  CREATE_TODO_SUCCESS,
  props<{
		todo: Todo;
	}>()
);

export const createTodoFailed = createAction(
  CREATE_TODO_FAILED,
  props<{
		error: string;
	}>()
);

export const deleteTodo = createAction(
  DELETE_TODO,
  props<{
		todoId: string;
	}>()
);

export const deleteTodoSuccess = createAction(
  DELETE_TODO_SUCCESS,
  props<{
		todoId: string;
	}>()
);

export const deleteTodoFailed = createAction(
  DELETE_TODO_FAILED,
  props<{
		error: string;
	}>()
);

export const updateTodo = createAction(
  UPDATE_TODO,
  props<{
		todoId: string;
		todo: UpdateTodo;
	}>()
);

export const updateTodoSuccess = createAction(
  UPDATE_TODO_SUCCESS,
  props<{
    todoId: string;
		todo: Todo;
	}>()
);

export const updateTodoFailed = createAction(
  UPDATE_TODO_FAILED,
  props<{
		error: string;
	}>()
);
