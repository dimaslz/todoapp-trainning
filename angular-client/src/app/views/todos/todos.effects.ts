import { NotificationService } from "@/app/components/notifications/notification.service";
import { CreateTodo, Todo, UpdateTodo } from "@/interfaces";
import { Injectable } from "@angular/core";
import { Actions, createEffect, ofType } from "@ngrx/effects";
import { of } from "rxjs";
import { map, mergeMap, catchError } from "rxjs/operators";

import { ApiService } from "../../services/api/api.service";

import {
  getTodos,
  getTodosSuccess,
  getTodosFailed,

  updateTodo,
  updateTodoSuccess,
  updateTodoFailed,

  createTodo,
  createTodoSuccess,
  createTodoFailed,

  deleteTodo,
  deleteTodoSuccess,
  deleteTodoFailed,
} from "./todos.actions";

@Injectable()
export class TodosEffects {

	getTodos$ = createEffect(() => this.actions$.pipe(
	  ofType(getTodos),
	  mergeMap(() => this.apiService.todos.getAll()
	      .pipe(
	        map(consult => getTodosSuccess({ data: consult })),
	        catchError(({ error }) => of(getTodosFailed({ error })))
	      ))
	));

	createTodo$ = createEffect(() => this.actions$.pipe(
	  ofType(createTodo),
	  mergeMap(({ todo }: { todo: CreateTodo }) => (
	    this.apiService.todos.create(todo)
	      .pipe(
	        map((data: Todo) => {
	          this.notificationsService.sendNotification("Task created!", { color: "green" });
	          return createTodoSuccess({ todo: data });
	        }),
	        catchError((error) => {
	          this.notificationsService.sendNotification(error);
	          return of(createTodoFailed({ error }));
	        })
	      ))
	  )
	));

	updateTodo$ = createEffect(() => this.actions$.pipe(
	  ofType(updateTodo),
	  mergeMap(({ todoId, todo }: { todoId: string; todo: UpdateTodo }) => (
	    this.apiService.todos.updateByID(todoId, todo)
	      .pipe(
	        map((data: Todo) => updateTodoSuccess({ todoId, todo: data })),
	        catchError((error) => of(updateTodoFailed({ error })))
	      ))
	  )
	));

	deleteTodo$ = createEffect(() => this.actions$.pipe(
	  ofType(deleteTodo),
	  mergeMap(({ todoId }: { todoId: string }) => (
	    this.apiService.todos.deleteByID(todoId)
	      .pipe(
	        map(() => deleteTodoSuccess({ todoId })),
	        catchError((error) => of(deleteTodoFailed({ error })))
	      ))
	  )
	));

	constructor(
		private actions$: Actions,
		private apiService: ApiService,
		private notificationsService: NotificationService,
	) {}
}
