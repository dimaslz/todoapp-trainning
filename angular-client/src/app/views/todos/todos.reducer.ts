/* eslint-disable no-underscore-dangle */
import { CreateTodo, Todo } from "@/interfaces";
import { createReducer, on } from "@ngrx/store";

import {
  getTodos,
  getTodosFailed,
  getTodosSuccess,

  createTodo,
  createTodoFailed,
  createTodoSuccess,

  deleteTodo,
  deleteTodoSuccess,
  deleteTodoFailed,

  updateTodo,
  updateTodoSuccess,
  updateTodoFailed,
} from "./todos.actions";

export type Status = {
	loading: boolean;
  loaded: boolean;
  error: string | string[] | boolean | null;
};

export type TodosState = {
  items: Todo[];
  status: Status;
};

export const INITIAL_STATE: TodosState = {
  items: [],
  status: {
    loading: true,
    loaded: false,
    error: null,
  }
};

export const reducer = createReducer(
  { ...INITIAL_STATE },
  on(createTodo, (state) => {
    state = {
      ...state,
      status: {
        ...state.status,
        loaded: false,
        loading: true,
        error: null,
      }
    };

    return state;
  }),
  on(createTodoSuccess, (state, { todo }: { todo: Todo }) => {
    state = {
      ...state,
      status: {
        ...state.status,
        loaded: true,
        loading: false,
      },
      items: state.items.concat([todo])
    };

    return state;
  }),
  on(createTodoFailed, (state, { error }: { error: string }) => {
    state = {
      ...state,
      status: {
        ...state.status,
        loaded: true,
        loading: false,
        error,
      },
    };

    return state;
  }),

  on(getTodos, (state) => {
    state = {
      ...state,
      status: {
        ...state.status,
        loading: true,
      }
    };

    return state;
  }),
  on(getTodosSuccess, (state, { data }: { data: Todo[] }) => {
    state = {
      ...state,
      status: {
        ...state.status,
        loading: false,
      },
      items: data
    };

    return state;
  }),
  on(getTodosFailed, (state, { error }: any) => {
    state = {
      ...state,
      status: {
        ...state.status,
        loading: false,
        error,
      },
    };

    return state;
  }),

  on(updateTodo, (state) => {
    state = {
      ...state,
      status: {
        ...state.status,
        loading: true,
      }
    };

    return state;
  }),
  on(updateTodoSuccess, (state, { todoId, todo }: { todoId: string; todo: Todo }) => {
    state = {
      ...state,
      status: {
        ...state.status,
        loading: false,
      },
      items: state.items.map((item: Todo) => {
        if (item._id === todoId) {
          return todo;
        }

        return item;
      }),
    };

    return state;
  }),
  on(updateTodoFailed, (state, { error }: any) => {
    state = {
      ...state,
      status: {
        ...state.status,
        loading: false,
        error,
      },
    };

    return state;
  }),


  on(deleteTodo, (state) => {
    state = {
      ...state,
      status: {
        ...state.status,
        loading: true,
      }
    };

    return state;
  }),
  on(deleteTodoSuccess, (state, { todoId }: { todoId: string }) => {
    state = {
      ...state,
      status: {
        ...state.status,
        loading: false,
      },
      items: state.items.filter((item: Todo) => item._id !== todoId)
    };

    return state;
  }),
  on(deleteTodoFailed, (state, { error }: { error: string }) => {
    state = {
      ...state,
      status: {
        ...state.status,
        loading: false,
        error
      },
    };

    return state;
  }),
);
