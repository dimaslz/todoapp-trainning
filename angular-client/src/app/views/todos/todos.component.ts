import { CreateTodo, Todo } from '@/interfaces';
import { Component, OnDestroy, OnInit } from '@angular/core';
import { FormControl, FormGroup, Validators } from '@angular/forms';
import { select, Store } from '@ngrx/store';
import { Subscription } from 'rxjs';
import { filter } from 'rxjs/operators';

import * as Actions from './todos.actions';

@Component({
  selector: 'app-todos',
  templateUrl: 'todos.component.html',
  styleUrls: ['./todos.component.scss']
})
export class TodosComponent implements OnInit, OnDestroy {
  public loading = true;
  public canCreate = false;
  public form = new FormGroup({
    title: new FormControl("", Validators.required),
  });

  public todos: Todo[] = [];

  private formSubscription: Subscription | null = null;
  private storeSubscription: Subscription | null = null;

  constructor(
    private store$: Store<any>,
  ) { }

  ngOnInit(): void {
    this.getTodos();

    this.formSubscription = this.form.valueChanges.subscribe(({ title }) => {
      this.canCreate = !!title;
    });

    this.storeSubscription = this.store$.pipe(
      select('todos'),
      filter(({ status }: any) => !status.loading),
      // filter(({ status }: any) => status.loaded)
    ).subscribe(({ items, status }: { items: Todo[]; status: any }) => {
      this.loading = status.loading;
      this.todos = items;
    });
  }

  ngOnDestroy(): void {
    this.storeSubscription?.unsubscribe();
    this.formSubscription?.unsubscribe();
  }

  createTodo(todo: CreateTodo): void {
    this.store$.dispatch(
      Actions.createTodo({
        todo,
      })
    );

    this.form.reset();
  }

  deleteTodo(todoId: string): void {
    this.store$.dispatch(
      Actions.deleteTodo({
        todoId,
      })
    );

    this.form.reset();
  }

  getTodos(): void {
    this.store$.dispatch(
      Actions.getTodos(),
    );
  }

  onSubmit(): void {
    const { title } = this.form.controls;

    const todo: CreateTodo = {
      title: title.value
    };

    this.createTodo(todo);
  }

  getTodoID(index: number, todo: any): any {
    return todo._id;
  }
}
