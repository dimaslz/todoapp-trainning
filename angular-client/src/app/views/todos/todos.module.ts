import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { RouterModule } from '@angular/router';
import { EffectsModule } from '@ngrx/effects';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { StoreModule } from '@ngrx/store';
import { NgCrossUIModule } from '@cross-ui/angular';

import { ApiService } from '@/app/services/api/api.service';

import { TodosRoutingModule } from './todos.routing.module';
import { TodosComponent } from './todos.component';
import { TodosEffects } from './todos.effects';

import { reducer } from './todos.reducer';
import { IconsModule } from '@/app/components/icons/icons.module';

@NgModule({
  declarations: [TodosComponent],
  imports: [
    CommonModule,
    FormsModule,
    ReactiveFormsModule,
    RouterModule,
    NgCrossUIModule,

    StoreModule.forFeature("todos", reducer),

    EffectsModule.forFeature([
      TodosEffects,
    ]),

    TodosRoutingModule,

    IconsModule,
  ],
  providers: [
    ApiService
  ]
})
export class TodosModule { }
