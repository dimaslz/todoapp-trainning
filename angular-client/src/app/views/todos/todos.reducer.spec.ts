import { CreateTodo, Todo } from '@/interfaces';
import * as Actions from './todos.actions';
import { reducer } from './todos.reducer';

describe('Todos reducer', () => {
  describe("get todos", () => {
    test('On -> getTodos', () => {
      const action = Actions.getTodos();
      const result = reducer({
        items: [],
        status: {
          loading: true,
          loaded: false,
          error: false,
        }
      }, action);

      expect(result).toMatchSnapshot();
    });

    test('On -> getTodosSuccess', () => {
      const todos: Todo[] = [{
        _id: "",
        title: "",
        description: "",
        status: {
          _id: "",
          name: "",
        },
      }];

      const action = Actions.getTodosSuccess({
        data: todos
      });

      const result = reducer({
        items: todos,
        status: {
          loading: false,
          loaded: true,
          error: false,
        }
      }, action);

      expect(result).toMatchSnapshot();
    });

    test('On -> getTodosFailed', () => {
      const error = "Something happen";

      const action = Actions.getTodosFailed({ error });

      const result = reducer({
        items: [],
        status: {
          loading: false,
          loaded: true,
          error,
        }
      }, action);

      expect(result).toMatchSnapshot();
    });
  });

  describe("create todo", () => {
    test('On -> createTodo', () => {
      const todo: CreateTodo = {
        title: "foo",
      };

      const action = Actions.createTodo({ todo });
      const result = reducer({
        items: [],
        status: {
          loading: true,
          loaded: false,
          error: false,
        }
      }, action);

      expect(result).toMatchSnapshot();
    });

    test('On -> createTodoSuccess', () => {
      const todo: Todo = {
        _id: "",
        title: "",
        description: "",
        status: {
          _id: "",
          name: "",
        },
      };

      const action = Actions.createTodoSuccess({
        todo,
      });

      const result = reducer({
        items: [todo],
        status: {
          loading: false,
          loaded: true,
          error: false,
        }
      }, action);

      expect(result).toMatchSnapshot();
    });

    test('On -> createTodoFailed', () => {
      const error = "Something happen";

      const action = Actions.createTodoFailed({ error });

      const result = reducer({
        items: [],
        status: {
          loading: false,
          loaded: true,
          error,
        }
      }, action);

      expect(result).toMatchSnapshot();
    });
  });

  describe("delete todo", () => {
    test('On -> deleteTodo', () => {
      const action = Actions.deleteTodo({ todoId: "todoId" });
      const result = reducer({
        items: [{
          _id: "todoId",
          title: "",
          description: "",
          status: {
            _id: "",
            name: "",
          },
        }],
        status: {
          loading: true,
          loaded: false,
          error: false,
        }
      }, action);

      expect(result).toMatchSnapshot();
    });

    test('On -> deleteTodoSuccess', () => {
      const action = Actions.deleteTodoSuccess({ todoId: "todoId" });
      const result = reducer({
        items: [{
          _id: "todoId",
          title: "",
          description: "",
          status: {
            _id: "",
            name: "",
          },
        }],
        status: {
          loading: true,
          loaded: false,
          error: false,
        }
      }, action);

      expect(result.items).toMatchObject([]);
    });

    test('On -> deleteTodoFailed', () => {
      const error = "error!";
      const action = Actions.deleteTodoFailed({ error });
      const items = [{
        _id: "todoId",
        title: "",
        description: "",
        status: {
          _id: "",
          name: "",
        },
      }];
      const result = reducer({
        items,
        status: {
          loading: false,
          loaded: true,
          error: false,
        }
      }, action);

      expect(result.items).toMatchObject(items);
      expect(result.status.error).toBe(error);
    });
  });

  describe("update todo", () => {
    test('On -> deleteTodo', () => {
      const action = Actions.updateTodo({
        todoId: "todoId",
        todo: {
          title: "new title",
        }
      });

      const result = reducer({
        items: [{
          _id: "todoId",
          title: "",
          description: "",
          status: {
            _id: "",
            name: "",
          },
        }],
        status: {
          loading: true,
          loaded: false,
          error: false,
        }
      }, action);

      expect(result).toMatchSnapshot();
    });

    test('On -> updateTodoSuccess', () => {
      const todo: Todo = {
        _id: "todoId",
        title: "new title",
        description: "",
        status: {
          _id: "",
          name: "",
        },
      };

      const action = Actions.updateTodoSuccess({
        todoId: "todoId",
        todo
      });
      const result = reducer({
        items: [{
          _id: "todoId",
          title: "",
          description: "",
          status: {
            _id: "",
            name: "",
          },
        }],
        status: {
          loading: false,
          loaded: true,
          error: false,
        }
      }, action);

      expect(result.items[0]).toMatchObject(todo);
    });

    test('On -> updateTodoFailed', () => {
      const error = "error!";
      const action = Actions.updateTodoFailed({ error });
      const items = [{
        _id: "todoId",
        title: "",
        description: "",
        status: {
          _id: "",
          name: "",
        },
      }];
      const result = reducer({
        items,
        status: {
          loading: false,
          loaded: true,
          error: false,
        }
      }, action);

      expect(result.items).toMatchObject(items);
      expect(result.status.error).toBe(error);
    });
  });
});
