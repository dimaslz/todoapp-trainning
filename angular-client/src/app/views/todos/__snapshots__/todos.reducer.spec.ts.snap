// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Todos reducer create todo On -> createTodo 1`] = `
Object {
  "items": Array [],
  "status": Object {
    "error": null,
    "loaded": false,
    "loading": true,
  },
}
`;

exports[`Todos reducer create todo On -> createTodoFailed 1`] = `
Object {
  "items": Array [],
  "status": Object {
    "error": "Something happen",
    "loaded": true,
    "loading": false,
  },
}
`;

exports[`Todos reducer create todo On -> createTodoSuccess 1`] = `
Object {
  "items": Array [
    Object {
      "_id": "",
      "description": "",
      "status": Object {
        "_id": "",
        "name": "",
      },
      "title": "",
    },
    Object {
      "_id": "",
      "description": "",
      "status": Object {
        "_id": "",
        "name": "",
      },
      "title": "",
    },
  ],
  "status": Object {
    "error": false,
    "loaded": true,
    "loading": false,
  },
}
`;

exports[`Todos reducer delete todo On -> deleteTodo 1`] = `
Object {
  "items": Array [
    Object {
      "_id": "todoId",
      "description": "",
      "status": Object {
        "_id": "",
        "name": "",
      },
      "title": "",
    },
  ],
  "status": Object {
    "error": false,
    "loaded": false,
    "loading": true,
  },
}
`;

exports[`Todos reducer get todos On -> getTodos 1`] = `
Object {
  "items": Array [],
  "status": Object {
    "error": false,
    "loaded": false,
    "loading": true,
  },
}
`;

exports[`Todos reducer get todos On -> getTodosFailed 1`] = `
Object {
  "items": Array [],
  "status": Object {
    "error": "Something happen",
    "loaded": true,
    "loading": false,
  },
}
`;

exports[`Todos reducer get todos On -> getTodosSuccess 1`] = `
Object {
  "items": Array [
    Object {
      "_id": "",
      "description": "",
      "status": Object {
        "_id": "",
        "name": "",
      },
      "title": "",
    },
  ],
  "status": Object {
    "error": false,
    "loaded": true,
    "loading": false,
  },
}
`;

exports[`Todos reducer update todo On -> deleteTodo 1`] = `
Object {
  "items": Array [
    Object {
      "_id": "todoId",
      "description": "",
      "status": Object {
        "_id": "",
        "name": "",
      },
      "title": "",
    },
  ],
  "status": Object {
    "error": false,
    "loaded": false,
    "loading": true,
  },
}
`;
